interfaces.go

```diff
@@ -79,6 +79,9 @@ type Method interface {
 
 	// Stores the next location to evaluate in xNext
 	Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
+
+	// Requires returns the minimum requirements of Method on Function.
```

Comment:
> On Mar 11, 2015, at 11:22 PM, Vladimír Chalupecký <notifications@github.com> wrote:
> 
> In interfaces.go <https://github.com/gonum/optimize/pull/84#discussion_r26280947>:
> 
> > @@ -79,6 +79,9 @@ type Method interface {
> >  
> >  	// Stores the next location to evaluate in xNext
> >  	Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
> > +
> > +	// Requires returns the minimum requirements of Method on Function.
> My intention for Requires() was to just provide information what kind of optimizer the Method is (derivative-free, gradient-based, ...), without considering how this information is used and how it is translated into interfaces that Function should satisfy. That's why I kept the comment rather minimal, because Requires() is of interest only to implementers of external minimizers and to us, not to users of optimize.
> 
The way the package is designed is that “users of optimize” is both those who want their functions minimized, and who want to use the “framework” of local when implementing a new optimizer. This is why Method is a publicly exported interface instead of it being a list of possible optimizers.  The Method documentation should thus speak to those who want to implement an algorithm.

You are right about a decoupling from the “interfaces that Function should satisfy”. Mostly, I thought the comment was a little spartan. When I get a chance I’ll try to say what I mean taking your comments into consideration.
> On the other hand, matching the requirements of Method and capabilites of Function is tricky and should be clearly explained in the documentation because it directly affects the users of optimize, it tells them what interfaces Function should satisfy. But Requires() is not the right place for such explanation, I would say that Local() or the package docs would be a better place.
> 
Agreed on both accounts
> I think that evaluations returned by Init() and Iterate() are more important for the matching between requirements and the methods of Function, maybe each optimizer should indicate the evaluations it requests in its docs.
> 

Interesting point. I’ll ponder.
> I will try to improve the comment for Requires() anyway.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/gonum/optimize/pull/84/files#r26280947>.
>